### DISCUSSION TITLE
Google OAuth hd query parameter

### DISCUSSION THREAD

**Discussion by gacharles23** [pocketbase/pocketbase#2782] [Q&A] [answered]
Using Google OAuth, is there a way to direct the login account selection to a specific domain via Google’s `hd` query parameter? Basically the equivalent of this Supabase code.

```
const { data, error: authError } = await event.locals.supabase.auth.signInWithOAuth({
  provider: 'google',
  options: {
    redirectTo: `${event.url.origin}/auth/callback`,
    queryParams: {
      hd: 'mydomain.com'
    }
  }
})
``` 

Thanks.

**Comment by ganigeorgiev** [comment]
You can add or replace any query parameter from the provider sign url by overwriting the default `urlCallback`, something like:

```js
await pb.collection("users").authWithOAuth2({
    provider: "google",
    urlCallback: (signUrl) => {
        let url = new URL(signUrl);
 
        url.searchParams.set("hd", "mydomain.com");
        
        window.open(url.toString(), '_blank');
    }
})
```

**Reply by gacharles23** [reply]
That is terrific. Thank you for the prompt response. I'm still trying to get basic Google OAuth working ... right now I'm getting a `originalError: ReferenceError: EventSource is not defined` error on just this: 

`await locals.pb.collection('users').authWithOAuth2({ provider: 'google' });`

But I'll keep at it and hopefully figure out what's wrong.

**Reply by ganigeorgiev** [reply]
The "all-in-one" OAuth2 flow uses the realtime subscriptions under the hood, which when using the JS SDK currently depends on [`EventSource`](https://developer.mozilla.org/en-US/docs/Web/API/EventSource).

What environment you are targetting? `EventSource` is usually natively supported in browsers. For React Native you'll need to load a polyfill (_see the install notes in https://github.com/pocketbase/js-sdk#nodejs-via-npm_).

Note that the above "all-in-one" flow may not be suitable for your use case since it expects to be executed client-side.

For a more advanced OAuth2 flow, you could check the "Manual code exchange" guide in https://pocketbase.io/docs/authentication/#oauth2-integration.

**Reply by gacharles23** [reply]
I'm building a SvelteKit app. Everything works fine for email/password login. 

This code is within a +page.server.js file for my /login route. So yeah, not client-side. Not sure what else to try.

**Reply by ganigeorgiev** [reply, marked as answer]
I'm not really sure how to explain it better and a lot of users seems to have difficulties with the OAuth2 when using a SSR framework or a node backend middleware.

You'll need an EventSource polyfill for node environments if you want to use the OAuth2 realtime subscription, from the [JS SDK readme](https://github.com/pocketbase/js-sdk#nodejs-via-npm):

```js
// for server: npm install eventsource --save
import eventsource from 'eventsource';

// for React Native: npm install react-native-sse --save
import eventsource from "react-native-sse";

global.EventSource = eventsource;
```

But keep in mind that the above "all-in-one" OAuth2 flow expects to be executed on the client-side since it needs to open a url in the client browser somehow.

If you want to manage the OAuth2 server-side you'll need to implement the "Manual code exchange" flow. Please check the guide in https://pocketbase.io/docs/authentication/#oauth2-integration. The shown example is for the web, but the general idea is the same and you'll need:

1. Somewhere to show the "Login with ..." links in your app.
2. An endpoint to handle the OAuth2 redirect and perform the code exchange (aka. the actual OAuth2 authentication).

**Reply by gacharles23** [reply]
Yeah, I definitely fall into that category of users with difficulties with OAuth2 using SSR! Anyway, thanks for the tips - I'll dig into it further and see what I can figure out. Thanks again for the prompt responses!
